<a href="https://colab.research.google.com/github/root-happy/Reyna/blob/main/data_science_pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<br>
<br>
<div>
<img  style="float: left; padding-right: 100px; width: 350px" src="logai.png">
    </div>
    <br>
    <h2 align="center">LogAI</h2>
<hr>
<h4 align="center"><a href="https://t.co/IoL5C0LiPX">Zephania Reuben</a></h4>
<h4 align="center">September, 2025</h4>
<hr>
<h3 align="center">AfriAI Lab - ADVANCED AI SHORT COURSE TRAINING 2025
<br>
<hr>
DS3: PANDAS
</h3>
<hr>
<br>

#### What is Pandas?
  - **Pandas** is a Python package providing fast, flexible, and expressive data structures designed to make working with “relational” or “labeled” data both easy and intuitive. It aims to be the fundamental high-level building block for doing practical, real world data analysis in Python.
  
  - The two primary data structures of pandas, Series (1-dimensional) and DataFrame (2-dimensional), handle the vast majority of typical use cases in finance, statistics, social science, and many areas of engineering.
  - Data frames are tabular, meaning that they are based on rows and columns like you would see in a spreadsheet.
  - pandas is built on top of NumPy and is intended to integrate well within a scientific computing environment with many other 3rd party libraries.
  
#### Here are just a few of the things that pandas does well:
 - Easy handling of missing data (represented as NaN) in floating point as well as non-floating point data
 - Size mutability: columns can be inserted and deleted from DataFrame and higher dimensional objects
 - Automatic and explicit data alignment: objects can be explicitly aligned to a set of labels, or the user cansimply ignore the labels and let Series, DataFrame, etc. automatically align the data for you in computations
 - Powerful, flexible group by functionality to perform split-apply-combine operations on data sets, for both aggregating and transforming data
 - Make it easy to convert ragged, differently-indexed data in other Python and NumPy data structures into DataFrame objects
 - Intelligent label-based slicing, fancy indexing, and subsetting of large data sets
 - Intuitive merging and joining data sets
 - Flexible reshaping and pivoting of data sets


#### Pandas Installation
- conda environment
 conda install pandas
- Installing from PyPI
   - python -m pip install pandas
- **Installing pandas on Linux**
 - In the following table, we will present some of the common Linux distributions package names for Matplotlib and the tools we can use to install the package:

**Distribution**        |       **Package Name**
----------------------  |  -------------------------------
Debian or Ubuntu (And other Debian derivatives)                 |  <code>sudo apt-get install python3-pandas</code>
Fedora                                                          |  <code>sudo dnf install python3-pandas</code>
Red hat                                                         |  <code>sudo yum install python3-pandas</code>
Centos/RHEL                                                            |  <code>sudo dnf install python3-pandas</code>


 #### 1.Understanding a pandas DataFrame
  - a pandas DataFrame (in a Jupyter Notebook) appears to be nothing more than an ordinary table of data consisting of rows and columns. Hiding beneath the surface are the three components--the index, columns, and data (also known as values) that you must be aware of in order to maximize the DataFrame's full potential.
  - Analyze the labeled anatomy of the DataFrame:
  - **Note**
   - In this Notebook we will be using a **Titanic** dataset.A dataset about passengers in Titanic.

**The variables that describe the passengers are:**

- **PassengerId**: and id given to each traveller on the boat.
- **Pclass**: the passenger class. It has three possible values: 1,2,3.
- **The Name**: a word or set of words by which a person or thing is usually known.
- **The Sex**: males or females considered as separate groups.
- **The Age**: the number of years that someone has lived.
- **SibSp**: number of siblings and spouses traveling with the passenger.
- **Parch**: number of parents and children traveling with the passenger.
- **The ticket number**: a number (identifier) piece of paper that shows you have paid for a journey.
- **The ticket Fare**: amount paid for a ticket.
- **The cabin number**: a number for private room on a ship for a passenger.
- **The embarkation**: It has three possible values S,C,Q

- A DataFrame has two axes: a **vertical axis** (the index) and a **horizontal axis**(the columns). Pandas borrows convention from NumPy and uses the integers 0/1 as another way of referring to the vertical/horizontal axis.

In [ ]:
#Load library
import pandas as pd

In [ ]:
import seaborn as sns

In [13]:
df = sns.load_dataset('titanic')

df = sns.load_dataset('titanic')

In [14]:
df.head(10)

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0         0       3    male  22.0      1      0   7.2500        S   Third   
1         1       1  female  38.0      1      0  71.2833        C   First   
2         1       3  female  26.0      0      0   7.9250        S   Third   
3         1       1  female  35.0      1      0  53.1000        S   First   
4         0       3    male  35.0      0      0   8.0500        S   Third   
5         0       3    male   NaN      0      0   8.4583        Q   Third   
6         0       1    male  54.0      0      0  51.8625        S   First   
7         0       3    male   2.0      3      1  21.0750        S   Third   
8         1       3  female  27.0      0      2  11.1333        S   Third   
9         1       2  female  14.0      1      0  30.0708        C  Second   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True  
5    man        True  NaN   Queenstown    no   True  
6    man        True    E  Southampton    no   True  
7  child       False  NaN  Southampton    no  False  
8  woman       False  NaN  Southampton   yes  False  
9  child       False  NaN    Cherbourg   yes  False

In [66]:
from google.colab import drive
drive.mount('/content/drive')


Mounted at /content/drive


In [69]:
#Create url

url = '/content/D1/ds/data/Titanic.csv'

In [68]:
# !cp -r /content/drive/MyDrive/D1 .

In [53]:
df.head(5)

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [72]:
# Load data as a DataFrame
dataframe = pd.read_csv(url)

In [73]:
dataframe.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q  
6      0             17463  51.8625   E46        S  
7      1            349909  21.0750   NaN        S  
8      2            347742  11.1333   NaN        S  
9      0            237736  30.0708   NaN        C

#### Things to notice in this DataFrame

- First, in a data frame each row corresponds to one observation (e.g., a passenger) and each column corresponds to one feature (gender, age, etc.). For example, bylooking at the first observation we can see that **Heikkinen, Miss. Laina** stayed in first class, was 26 years old, was female, and survived the disaster.
- Second, each column contains a name (e.g., Name, PClass, Age) and each rowcontains an index number (e.g., 0 for the lucky Miss Elisabeth Walton Allen). We will use these to select and manipulate observations and features.

#### 2.Creating a DataFrame
 - First method :
   - Create a dataframe and add columns independently.

In [74]:
df = pd.DataFrame()

In [76]:
df['names'] = ['John','Rose','Jack']

In [77]:
df['age'] = [15,19,46]

In [78]:
df['country'] = ['Kenya','Uganda','Malawi']

In [79]:
df

names  age country
0  John   15   Kenya
1  Rose   19  Uganda
2  Jack   46  Malawi

- Second method :
   - Create a dataframe and add columns at the same time.

In [80]:
#Load library
import pandas as pd

# Create a DataFrame
df = pd.DataFrame(columns=['Name','Age','Country'],
                  data=[
                       ['John',19,'Kenya'],
                       ['Rebecca',16,'Uganda'],
                       ['Lisa',19,'Rwanda'],
                       ['Godfrey',19,'Tanzania'],
                       ['Vivian',19,'Burundi']
                      ])

#show DataFrame
df

Name  Age   Country
0     John   19     Kenya
1  Rebecca   16    Uganda
2     Lisa   19    Rwanda
3  Godfrey   19  Tanzania
4   Vivian   19   Burundi

In [81]:
df.iloc[0]

Name        John
Age           19
Country    Kenya
Name: 0, dtype: object

In [82]:
type(df)

pandas.core.frame.DataFrame

#### 3.Creating a Series

In [83]:
#Load library
import pandas as pd

#Create a Series
series = pd.Series(index=['Name','Age','Country'],data=['John',19,'Uganda'])

#show series
series

Name         John
Age            19
Country    Uganda
dtype: object

In [ ]:
type(series)

#### A series can be used to create a DataFrame as follows

In [86]:
df = pd.DataFrame()

In [87]:
df

Empty DataFrame
Columns: []
Index: []

#### 4.Describing a DataFrame
- Describing a DataFrame involve looking at its short summary of descriptive statistical measures.

In [88]:
dataframe.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

- We can also take a look at the number of row and colums

In [89]:
dataframe.shape

(891, 12)

In [90]:
dataframe.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


- DataFrame has 891 rows(instances/samples) and 12 colums(features)

#### 5.Navigating DataFrames
  - You need to select individual data or slices of a DataFrame
    - **loc**
      - is useful when the index of the DataFrame is a label (e.g., a string).
    - **iloc**
      - works by looking for the position in the DataFrame. For example, iloc[0] will return the first row regardless of whether the index is an integer or a label.

In [91]:
dataframe.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [92]:
# Select three rows
dataframe.iloc[1:4] # also dataframe.iloc[:4]

PassengerId  Survived  Pclass  \
1            2         1       1   
2            3         1       3   
3            4         1       1   

                                                Name     Sex   Age  SibSp  \
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   

   Parch            Ticket     Fare Cabin Embarked  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S

- DataFrames do not need to be numerically indexed. We can set the index of a DataFrame to any value where the value is unique to each row. For example, we can set the index to be passenger names and then select rows using a name:

In [94]:
#set index

dataframe = dataframe.set_index(dataframe['Name'])

In [95]:
dataframe.head()

PassengerId  Survived  \
Name                                                                        
Braund, Mr. Owen Harris                                       1         0   
Cumings, Mrs. John Bradley (Florence Briggs Tha...            2         1   
Heikkinen, Miss. Laina                                        3         1   
Futrelle, Mrs. Jacques Heath (Lily May Peel)                  4         1   
Allen, Mr. William Henry                                      5         0   

                                                    Pclass  \
Name                                                         
Braund, Mr. Owen Harris                                  3   
Cumings, Mrs. John Bradley (Florence Briggs Tha...       1   
Heikkinen, Miss. Laina                                   3   
Futrelle, Mrs. Jacques Heath (Lily May Peel)             1   
Allen, Mr. William Henry                                 3   

                                                                                                 Name  \
Name                                                                                                    
Braund, Mr. Owen Harris                                                       Braund, Mr. Owen Harris   
Cumings, Mrs. John Bradley (Florence Briggs Tha...  Cumings, Mrs. John Bradley (Florence Briggs Th...   
Heikkinen, Miss. Laina                                                         Heikkinen, Miss. Laina   
Futrelle, Mrs. Jacques Heath (Lily May Peel)             Futrelle, Mrs. Jacques Heath (Lily May Peel)   
Allen, Mr. William Henry                                                     Allen, Mr. William Henry   

                                                       Sex   Age  SibSp  \
Name                                                                      
Braund, Mr. Owen Harris                               male  22.0      1   
Cumings, Mrs. John Bradley (Florence Briggs Tha...  female  38.0      1   
Heikkinen, Miss. Laina                              female  26.0      0   
Futrelle, Mrs. Jacques Heath (Lily May Peel)        female  35.0      1   
Allen, Mr. William Henry                              male  35.0      0   

                                                    Parch            Ticket  \
Name                                                                          
Braund, Mr. Owen Harris                                 0         A/5 21171   
Cumings, Mrs. John Bradley (Florence Briggs Tha...      0          PC 17599   
Heikkinen, Miss. Laina                                  0  STON/O2. 3101282   
Futrelle, Mrs. Jacques Heath (Lily May Peel)            0            113803   
Allen, Mr. William Henry                                0            373450   

                                                       Fare Cabin Embarked  
Name                                                                        
Braund, Mr. Owen Harris                              7.2500   NaN        S  
Cumings, Mrs. John Bradley (Florence Briggs Tha...  71.2833   C85        C  
Heikkinen, Miss. Laina                               7.9250   NaN        S  
Futrelle, Mrs. Jacques Heath (Lily May Peel)        53.1000  C123        S  
Allen, Mr. William Henry                             8.0500   NaN        S

In [96]:
#use index to slice and show row
dataframe.loc['Heikkinen, Miss. Laina']

PassengerId                         3
Survived                            1
Pclass                              3
Name           Heikkinen, Miss. Laina
Sex                            female
Age                              26.0
SibSp                               0
Parch                               0
Ticket               STON/O2. 3101282
Fare                            7.925
Cabin                             NaN
Embarked                            S
Name: Heikkinen, Miss. Laina, dtype: object

#### 6.Selecting Rows Based on Conditionals
 - Suppose we want to select all women in Titanic

In [97]:
# Load library
import pandas as pd

# Create URL
url = './Titanic.csv'

# Load data
dataframe = pd.read_csv(url)

# Show top two rows where column 'sex' is 'female'
dataframe[ dataframe['Sex']=='female' ].head(2)

PassengerId  Survived  Pclass  \
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S

- Multiple conditions are easy as well. For example, here we select all the rows where the passenger is a female 65 or older:


In [ ]:
# Show top two rows where column 'sex' is 'female' and 'age' >=27
dataframe[(dataframe['Sex'] == 'female') & (dataframe['Age'] >= 27) ].head(10)

#### 7.Replacing Values
 - pandas’ replace is an easy way to find and replace values. For example, we can replace any instance of "female" in the Sex column with "Woman":

In [98]:
# Load library
import pandas as pd

# Create URL
url = './Titanic.csv'

# Load data
dataframe = pd.read_csv(url)

# Replace values, show two rows
dataframe[dataframe['Sex']=='female'].replace("female", "F").head(10)

PassengerId  Survived  Pclass  \
1             2         1       1   
2             3         1       3   
3             4         1       1   
8             9         1       3   
9            10         1       2   
10           11         1       3   
11           12         1       1   
14           15         0       3   
15           16         1       2   
18           19         0       3   

                                                 Name Sex   Age  SibSp  Parch  \
1   Cumings, Mrs. John Bradley (Florence Briggs Th...   F  38.0      1      0   
2                              Heikkinen, Miss. Laina   F  26.0      0      0   
3        Futrelle, Mrs. Jacques Heath (Lily May Peel)   F  35.0      1      0   
8   Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)   F  27.0      0      2   
9                 Nasser, Mrs. Nicholas (Adele Achem)   F  14.0      1      0   
10                    Sandstrom, Miss. Marguerite Rut   F   4.0      1      1   
11                           Bonnell, Miss. Elizabeth   F  58.0      0      0   
14               Vestrom, Miss. Hulda Amanda Adolfina   F  14.0      0      0   
15                   Hewlett, Mrs. (Mary D Kingcome)    F  55.0      0      0   
18  Vander Planke, Mrs. Julius (Emelia Maria Vande...   F  31.0      1      0   

              Ticket     Fare Cabin Embarked  
1           PC 17599  71.2833   C85        C  
2   STON/O2. 3101282   7.9250   NaN        S  
3             113803  53.1000  C123        S  
8             347742  11.1333   NaN        S  
9             237736  30.0708   NaN        C  
10           PP 9549  16.7000    G6        S  
11            113783  26.5500  C103        S  
14            350406   7.8542   NaN        S  
15            248706  16.0000   NaN        S  
18            345763  18.0000   NaN        S

- We can also replace multiple values at the same time:

In [99]:

# Load library
import pandas as pd

# Create URL
url = './Titanic.csv'

# Load data
dataframe = pd.read_csv(url)

# Replace "female" and "male with "Woman" and "Man"
dataframe[(dataframe['Sex'] == 'female') | (dataframe['Sex'] == 'male')].replace(["female", "male"], [0, 1]).head(5)

/tmp/ipython-input-4225231867.py:11: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  dataframe[(dataframe['Sex'] == 'female') | (dataframe['Sex'] == 'male')].replace(["female", "male"], [0, 1]).head(5)


PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name  Sex   Age  SibSp  Parch  \
0                            Braund, Mr. Owen Harris    1  22.0      1      0   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...    0  38.0      1      0   
2                             Heikkinen, Miss. Laina    0  26.0      0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)    0  35.0      1      0   
4                           Allen, Mr. William Henry    1  35.0      0      0   

             Ticket     Fare Cabin Embarked  
0         A/5 21171   7.2500   NaN        S  
1          PC 17599  71.2833   C85        C  
2  STON/O2. 3101282   7.9250   NaN        S  
3            113803  53.1000  C123        S  
4            373450   8.0500   NaN        S

In [100]:
dataframe['Ticket'][1].split()[1]

'17599'

In [101]:
dataframe.Ticket[1].split()

['PC', '17599']

In [102]:
#Passenger ID

#### 8.Renaming Columns
- Rename columns using the rename method:

In [103]:
# Load library
import pandas as pd

# Create URL
url = './Titanic.csv'

# Load data
dataframe = pd.read_csv(url)

# Rename column, show two rows
dataframe.rename(columns={'Pclass': 'p_class'}).head(2)

PassengerId  Survived  p_class  \
0            1         0        3   
1            2         1        1   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   

   Parch     Ticket     Fare Cabin Embarked  
0      0  A/5 21171   7.2500   NaN        S  
1      0   PC 17599  71.2833   C85        C

- Notice that the rename method can accept a dictionary as a parameter. We can use the dictionary to change multiple column names at once:

In [ ]:
# Rename columns, show two rows
dataframe.rename(columns={'Pclass': 'p_class', 'Sex': 'sex'}).head(2)

#### 9.Finding the Minimum, Maximum, Sum,Average, and Count

In [104]:
# Load library
import pandas as pd
# Create URL
url = './Titanic.csv'
# Load data
dataframe = pd.read_csv(url)
# Calculate statistics
print('Maximum:', dataframe['Age'].max())
print('Minimum:', dataframe['Age'].min())
print('Mean:', dataframe['Age'].mean())
print('Sum:', dataframe['Age'].sum())
print('Count:', dataframe['Age'].count())

Maximum: 80.0
Minimum: 0.42
Mean: 29.69911764705882
Sum: 21205.17
Count: 714


In [106]:
dataframe['Age'].agg('mean')

np.float64(29.69911764705882)

In [105]:
dataframe['Age'].agg('median')

28.0

In [135]:
dataFrame['Age'].agg('mode')

0    24.0
Name: Age, dtype: float64

#### 10.Finding Unique Values
 - Use unique to view an array of all unique values in a column:

In [136]:
# Load library
import pandas as pd

# Create URL
url = './Titanic.csv'

# Load data
dataframe = pd.read_csv(url)

# Select unique values
dataframe['Pclass'].unique()

array([3, 1, 2])

In [137]:
dataframe['Pclass'].value_counts()

Pclass
3    491
1    216
2    184
Name: count, dtype: int64

- Alternatively, value_counts will display all unique values with the number of times each value appears:

In [138]:
dataframe['Sex'].value_counts()

Sex
male      577
female    314
Name: count, dtype: int64

#### 11.Handling Missing Values
 - isnull and notnull return booleans indicating whether a value is missing:

In [139]:
# Load library
import pandas as pd

# Create URL
url = './Titanic.csv'

# Load data
dataframe = pd.read_csv(url)

## Select missing values, show two rows
dataframe[dataframe['Age'].isnull()].head(2)

PassengerId  Survived  Pclass                          Name   Sex  Age  \
5             6         0       3              Moran, Mr. James  male  NaN   
17           18         1       2  Williams, Mr. Charles Eugene  male  NaN   

    SibSp  Parch  Ticket     Fare Cabin Embarked  
5       0      0  330877   8.4583   NaN        Q  
17      0      0  244373  13.0000   NaN        S

In [114]:
(dataframe.isnull().sum()/dataframe.shape[0])*100

PassengerId     0.000000
Survived        0.000000
Pclass          0.000000
Name            0.000000
Sex             0.000000
Age            19.865320
SibSp           0.000000
Parch           0.000000
Ticket          0.000000
Fare            0.000000
Cabin          77.104377
Embarked        0.224467
Age_filled      0.000000
dtype: float64

In [145]:
dataFrame['Age_filled'] = dataFrame['Age'].fillna(value=dataFrame['Age'].median())

In [147]:
dataFrame['Age_filled'].isna().sum()

np.int64(0)

In [128]:
dataframe.shape

(891, 13)

#### 12.Deleting a Column

 - The best way to delete a column is to use drop with the parameter axis=1 (i.e., the column axis):

In [148]:
# Load library
import pandas as pd

# Create URL
url = './Titanic.csv'

# Load data
dataframe = pd.read_csv(url)

# Delete column
dataframe = dataframe.drop('Age', axis=1).head(2)

In [130]:
dataframe.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'SibSp', 'Parch',
       'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

- You can also use a list of column names as the main argument to drop multiple columns at once:

In [149]:
# Drop columns
dataframe.drop(['Pclass', 'Sex'], axis=1)

PassengerId  Survived                                               Name  \
0            1         0                            Braund, Mr. Owen Harris   
1            2         1  Cumings, Mrs. John Bradley (Florence Briggs Th...   

   SibSp  Parch     Ticket     Fare Cabin Embarked  
0      1      0  A/5 21171   7.2500   NaN        S  
1      1      0   PC 17599  71.2833   C85        C

In [ ]:
[dataframe['PassengerId'] != 2].head()

#### 13.Deleting a Row
 - Use a boolean condition to create a new DataFrame excluding the rows you want to delete:

In [150]:
# Load library
import pandas as pd

# Create URL
url = './Titanic.csv'

# Load data
dataframe = pd.read_csv(url)

# Delete rows, show first two rows of output
dataframe[dataframe['PassengerId'] != 3].head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
3            4         1       1   
4            5         0       3   
5            6         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   

   Parch     Ticket     Fare Cabin Embarked  
0      0  A/5 21171   7.2500   NaN        S  
1      0   PC 17599  71.2833   C85        C  
3      0     113803  53.1000  C123        S  
4      0     373450   8.0500   NaN        S  
5      0     330877   8.4583   NaN        Q

#### 14.Dropping Duplicate Rows
- Use drop_duplicates, but be mindful of the parameters:

In [152]:
# Load library
import pandas as pd
# Create URL
url = './Titanic.csv'
# Load data
dataframe = pd.read_csv(url)

In [153]:
dataframe.duplicated().sum()

np.int64(0)

In [154]:
# Drop duplicates, show first two rows of output
dataframe.drop_duplicates(keep='first').head(2)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   

   Parch     Ticket     Fare Cabin Embarked  
0      0  A/5 21171   7.2500   NaN        S  
1      0   PC 17599  71.2833   C85        C

#### 15.Grouping Rows by Values
- groupby is one of the most powerful features in pandas:

In [155]:
# Load library
import pandas as pd

# Create URL
url = './Titanic.csv'

# Load data
dataframe = pd.read_csv(url)

# Group rows by the values of the column 'Sex', calculate mean
# of each group
dataframe[['Sex','Age']].groupby('Sex').mean().rename(columns={'Age':'mean_age'})

mean_age
Sex              
female  27.915709
male    30.726645

#### 15.Concatenating DataFrames
 - Use concat with axis=0 to concatenate along the row axis:

In [156]:
# Load library
import pandas as pd

# Create DataFrame
data_a = {'id': ['1', '2', '3'],
'first': ['Alex', 'Amy', 'Allen']}
dataframe_a = pd.DataFrame(data = data_a, columns = ['id', 'first'])


In [157]:
dataframe_a

id  first
0  1   Alex
1  2    Amy
2  3  Allen

In [163]:
# Create DataFrame
data_b = {'id': ['1', '2', '3','4'],'last': ['Anderson', 'Ackerman', 'Ali','Juma']}
dataframe_b = pd.DataFrame(data = data_b, columns = ['id', 'last'])

In [167]:
dataframe_b

id      last
0  1  Anderson
1  2  Ackerman
2  3       Ali
3  4      Juma

In [168]:
pd.merge(dataframe_a,dataframe_b, how='outer',on='id')

id  first      last
0  1   Alex  Anderson
1  2    Amy  Ackerman
2  3  Allen       Ali
3  4    NaN      Juma

In [169]:
pd.concat([dataframe_a,dataframe_b], axis=0)

id  first      last
0  1   Alex       NaN
1  2    Amy       NaN
2  3  Allen       NaN
0  1    NaN  Anderson
1  2    NaN  Ackerman
2  3    NaN       Ali
3  4    NaN      Juma

<h3 style="color:red;">Referencies</h3>

- Machine Learning with Python
Cookbook,Chris Albon, O’Reilly Media, Inc,2018